### PR TITLE
Optimize and parallelize density computation

### PR DESCRIPTION
**Pull request by antoine-levitt**

### CODE REVIEW

File: src/terms/Hamiltonian.jl [hunk 1]
```diff
@@ -87,7 +87,7 @@ Base.:*(H::Hamiltonian, ψ) = mul!(deepcopy(ψ), H, ψ)
 # (eg the density ρ)
 @timing function energy_hamiltonian(basis::PlaneWaveBasis, ψ, occ; kwargs...)
     # it: index into terms, ik: index into kpoints
-    ene_ops_arr = [ene_ops(term, ψ, occ; kwargs...) for term in basis.terms]
+    @timing "ene_ops" ene_ops_arr = [ene_ops(term, ψ, occ; kwargs...) for term in basis.terms]
```
Comment (mfherbst) [inline src/terms/Hamiltonian.jl:90]: Should be consistent with https://github.com/JuliaMolSim/DFTK.jl/blob/master/src/terms/local.jl#L98. One way or the other ... I don't care.

File: src/densities.jl [hunk 2]
```diff
@@ -26,7 +32,7 @@ function compute_partial_density(basis, kpt, ψk, occupation)
     end
 
     # FFT and return
-    r_to_G(basis, ρk_real)
+    r_to_G!(ρ, basis, complex.(ρk_real))
```
Comment (mfherbst) [inline src/densities.jl:35]: You don't need a `complex.` here just `complex` does it (either way a copy).

**Review by mfherbst — COMMENTED** [review]
Nice. ~~You forgot to update the LDOS routine.~~

**Update:** Ah no sorry that's a different layer.

File: src/terms/Hamiltonian.jl (hunk 1 above)
Comment (antoine-levitt) [reply src/terms/Hamiltonian.jl:90]: I don't understand what you mean here

File: src/densities.jl (hunk 2 above)
Comment (antoine-levitt) [reply src/densities.jl:35]: good catch

File: src/terms/Hamiltonian.jl (hunk 1 above)
Comment (mfherbst) [reply src/terms/Hamiltonian.jl:90]: Ah sorry ... wrong annotation. I meant the individual `ene_ops` functions for each term where you use a different convention than for the timing label to the respective force terms.

**Pull request merged**